instance/factories.py

```diff
@@ -0,0 +1,122 @@
+from django.template import loader
+
+from instance.models.openedx_instance import OpenEdXInstance
+
+
+# Logging #####################################################################
+
+logger = logging.getLogger(__name__)
+
+
+# Functions ###################################################################
+
+def _check_environment():
+    """
+    Check environment and report potential problems for production instances
+    """
+    if not settings.SWIFT_ENABLE:
+        logger.warning("Swift support is currently disabled. Adjust SWIFT_ENABLE setting.")
+    if settings.INSTANCE_MYSQL_URL is None:
+        logger.warning("URL for external MySQL database is missing. Adjust INSTANCE_MYSQL_URL setting.")
+    if settings.INSTANCE_MONGO_URL is None:
+        logger.warning("URL for external Mongo database is missing. Adjust INSTANCE_MONGO_URL setting.")
+
+
+def instance_factory(**kwargs):
+    """
+    Factory function for creating instances.
+
+    Returns a newly created OpenEdXInstance.
+
+    Callers can use keyword arguments to pass in non-default values
+    for any field that is defined on the OpenEdXInstance model.
+
+    The only mandatory argument is `sub_domain`.
+
+    When called without any additional arguments, the instance that is returned
+    will have its fields set to default values that are appropriate
+    for *sandbox* instances.
+
+    To create an instance with default settings that are suitable for production,
+    use `production_instance_factory`.
+    """
+    # Ensure caller provided required arguments
+    assert "sub_domain" in kwargs
+
+    # Ensure instance uses ephemeral databases
+    instance_kwargs = dict(use_ephemeral_databases=True)
+    instance_kwargs.update(kwargs)
+
+    # Create instance
+    instance = OpenEdXInstance.objects.create(**instance_kwargs)
+    return instance
+
+
+def production_instance_factory(**kwargs):
+    """
+    Factory function for creating production instances.
+
+    Returns a newly created OpenEdXInstance.
+
+    Callers can use keyword arguments to pass in non-default values
+    for any field that is defined on the OpenEdXInstance model.
+
+    The only mandatory argument is `sub_domain`.
+
+    When called without any additional arguments, the instance that is returned
+    will have its fields set to default values that are appropriate
+    for *production* instances.
+
+    To create an instance with default settings that are suitable for sandboxes,
+    use `instance_factory`.
+    """
+    # NOTE: The long-term goal is to eliminate differences between sandboxes
+    # and production instances, and for this function to disappear.
+    # Please do not add behavior that is specific to production instances here.
+
+    # Ensure caller provided required arguments
+    assert "sub_domain" in kwargs
+
+    # Check environment and report potential problems
+    _check_environment()
+
+    # Create instance
+    production_instance = OpenEdXInstance(**kwargs)
```

Comment:
Is there a reason you use `.objects.create()` for sandboxes but instantiate-update-save() for the production factory? I think it would be easier to compare the two methods if you took the same approach as above here, i.e. update `instance_kwargs` then just call `.create(**instance_kwargs)`